### PR TITLE
refactor: convert CsvMergedHeadersGenerator to typescript

### LINKED ISSUE #2030: Convert CsvMergedHeadersGenerator class and associated functions to TypeScript
To follow the parent `CsvGenerator` class, which was already converted to TS in #1899 . Do also look into migrating related code if it makes sense.

### PR DESCRIPTION
**Pull request by chowyiyin**
## Problem
This PR converts the `CsvMergedHeadersGenerator` to typescript.

Closes #2030

## Solution
Apart from converting the `CsvMergedHeadersGenerator` class to typescript, the response classes in `csv-response-classes` were also converted to typescript. Also, the `Response` class was converted to an abstract class, which cannot be instantiated. All subclass will have to implement the functions `getAnswer` and `numCols`. This was done as the existing implementation in `CsvMergedHeadersGenerator` implies the need for all subclasses of `Response` to have the two aforementioned functions.

## Manual Tests
- [ ] Create a storage mode form with a table field (to check `TableResponse`), a checkbox field (to check `ArrayAnswerResponse`) and an email field (to check `SingleAnswerResponse`). Submit a few responses and download the csv. Ensure that all answers to each form field are present in the csv and the metadata headers are also present.

### CODE REVIEW

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:1]: maintain the import of the specific library 'lodash/keyBy' so we minimise the bundle size of the frontend

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:47]: we should remove the JSdocs types, since typescript code already contains more specific types. no point maintaining the type documentation in two places

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:14]: haha we're going to have to be more specific than this - have you checked out the documentation for `keyBy`? if you actually hover your mouse over `ReturnType<typeof keyBy>`, you'll notice that it's `any` :P

File: src/public/modules/forms/helpers/csv-response-classes/TableResponse.class.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/csv-response-classes/TableResponse.class.ts:19]: `this._data` should be typed as a class member. I think that should work if you get its type to inherit from `DisplayedResponseWithoutAnswer`

File: src/public/modules/forms/helpers/csv-response-classes/SingleAnswerResponse.class.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/csv-response-classes/SingleAnswerResponse.class.ts:11]: again, this class member should be typed more specifically

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:141]: you don't actually need to specify these types. since `fieldIdToQuestion` is a typed Map, its `forEach` is also automatically typed.

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:105]: the problem with `Dictionary` is that it assumes that every string is a valid key. for example, if you do `const value = unprocessedRecord['some crap']`, `value` will be typed as a `Response`. so instead of using `Dictionary` it would be better to declare our own specific type.

**Review by mantariksh — CHANGES_REQUESTED** [review]
I think we should type `getResponseInstance` as part of this as well, wdyt?

File: src/public/modules/forms/helpers/csv-response-classes/Response.class.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/csv-response-classes/Response.class.ts:4]: By defining the type of `_data` as `DisplayedResponseWithoutAnswer` in `Response.class.ts`, `_data` is typed as `DisplayedResponseWithoutAnswer` and therefore doesn't have the properties `answerArray` or `answer`. Is there a way to get pass this problem?

Comment (mantariksh) [reply src/public/modules/forms/helpers/csv-response-classes/Response.class.ts:4]: does it work if you type the `_data` member more specifically in the subclasses? eg can you do `_data: DisplayedResponseWithoutAnswer` in `Response.class` and `_data: NestedArrayResponse` in `TableResponse.class`?

Comment (chowyiyin) [reply src/public/modules/forms/helpers/csv-response-classes/Response.class.ts:4]: oh yes it does thank you!

File: src/types/response/index.ts
Comment (mantariksh) [inline src/types/response/index.ts:74]: nit but let's stick to either `type` or `interface`, no point mixing the two.

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:39]: let's stick to arrow syntax instead of `function` syntax for style consistency

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:42]: this is not quite enough to determine that the response is a `NestedResponse`, right? we should also check that the array is in fact nested. the `hasProp` utility in the backend might come in useful here; you can consider moving that into the `shared` folder so the frontend can make use of it.

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:47]: similarly here we should check the type of `answerArray`, not just the presence of the key. e.g. what if it's undefined?

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:53]: same here - we expect `answer` to be of a specific type, in this case I think it's `string` (but maybe also `number`? you'll have to check)

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:36]: interesting approach! my concern here is that this will fail silently at the point where we actually discover the error, and then unexpectedly break something else down the line. I _think_ if you simply throw an error here, it will bubble up and get caught by `worker.on('error'` - could you help to verify this? if that's the case then I think we should rely on existing error handling mechanisms (i.e. `worker.on('error'` and `errorCount` in `submissions.client.factory`)

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:29]: this mixing of `switch` and `if` is a little confusing. I'd actually much rather stick to just if-else.

**Review by mantariksh — CHANGES_REQUESTED** [review]

File: src/public/modules/forms/helpers/response-factory.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/response-factory.ts:36]: I just tried and the error doesn't get caught in `worker.on('error'`. I think it will only get caught if the error occurred in the decryption worker. But I do agree this approach would be better so that its clearer that an error occurred. I'll just wrap the call to `addRecord` with a try catch block and increment `errorCount when there's an error.

File: src/public/modules/forms/helpers/response-factory.ts
Comment (seaerchin) [inline src/public/modules/forms/helpers/response-factory.ts:42]: wdyt about extracting 

```ts
hasProp(response, 'answerArray') &&
Array.isArray(response.answerArray)
```

into its own function (maybe `isAnswerArray`) and then having the additional subtyping checks?

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:105]: Correct me if I'm wrong but I think `#extractAnswer` is not allowed?

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:60]: Nope. The `record` in `UnprocessedRecord` is of another type: `{ [fieldId: string]: Response }`

Comment (seaerchin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:60]: Ah ok! Wdyt about making this a type alias for readability

File: src/public/modules/forms/helpers/response-factory.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/response-factory.ts:42]: Hm.. I know I won't be able to do this with a normal `isAnswerArray` function that returns a boolean as typescript will complain that `answerArray` may not be defined. Do you know of something similar to a type guard that can help to overcome this issue?

File: src/public/modules/forms/helpers/response-factory.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/response-factory.ts:16]: I did think about how to add more clarity to the type signature so that it would be clear that an error could be created here. But the benefit of the current approach is that the error can just bubble up to `submissions.client.factory` and it doesn't have to be handled directly in `addRecord`. I did add the comments to the relevant portions of the code to indicate the possibility of an error. Do you think this is sufficient?

File: src/public/modules/forms/helpers/response-factory.ts
Comment (seaerchin) [reply src/public/modules/forms/helpers/response-factory.ts:42]: i think you'd need to create a base type! so the base function is a type predicate for that then the other nested ones just extend the base type so ts can infer properly!

File: src/public/modules/forms/helpers/response-factory.ts
Comment (seaerchin) [reply src/public/modules/forms/helpers/response-factory.ts:16]: i think not handling here is ok but the comments have to be in the jsdoc, because otherwise they won't be reflected on mouse hover @__@ 

the main problem is that errors if not stated in the signature, should be plainly obvious to the caller! adding the comment is good but cos it's not in the JSDoc, callers won't see on hover, which defeats the purpose of the comment

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (seaerchin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:105]: i think you'd have to declare it in the class then initialize it in the constructor! refer to myinfo.service.ts but it should be possible!

File: src/public/modules/forms/helpers/response-factory.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/response-factory.ts:16]: Oh right... I'll add it in thank you!!

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:60]: I would prefer not to haha :) since this type is only referenced once

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts]: Not possible due to ECMAScript version

**Review by mantariksh — COMMENTED** [review]
pinging @karrui for another pair of eyes

File: tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js
Comment (chowyiyin) [reply tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js:176]: This test no longer passes with the new implementation because we can't instantiate a `Response` object, and any object that doesn't fit the three response types will result in an error. I'm not completely sure and do correct me if I'm wrong, but I don't think there's a case where we will try to do this?

Comment (mantariksh) [reply tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js:176]: I'm pretty sure this is important because it seems to be checking for the case where a form contains a single statement field. @karrui can correct me if I'm wrong. does the code work for that case?

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (karrui) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:61]: this should be a type. even though its only used once in this file, functions calling this may need the type.

for example, i see that in the current code this type is actually a CsvRecord.submissionData from the decryption worker (which has not been translated to Typescript yet),

Ideally, the type would be exported from the decryption worker and imported and used here, but since that does not exist yet we should type it here.

File: tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js
Comment (chowyiyin) [reply tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js:176]: I tested and it still works. I printed the record for statement fields and the `answer` has value `Success` so a `SingleAnswerResponse` object will still be created

File: src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts
Comment (karrui) [inline src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts:6]: can probably mark this as private with javascript's private modifier [`#`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields) and remove the underscore hehe

File: src/public/modules/forms/helpers/csv-response-classes/Response.class.ts
Comment (karrui) [inline src/public/modules/forms/helpers/csv-response-classes/Response.class.ts:4]: same here

File: src/public/modules/forms/helpers/csv-response-classes/SingleAnswerResponse.class.ts
Comment (karrui) [inline src/public/modules/forms/helpers/csv-response-classes/SingleAnswerResponse.class.ts:6]: same here

File: src/public/modules/forms/helpers/csv-response-classes/TableResponse.class.ts
Comment (karrui) [inline src/public/modules/forms/helpers/csv-response-classes/TableResponse.class.ts:6]: same here

File: src/public/modules/forms/services/submissions.client.factory.js
Comment (karrui) [inline src/public/modules/forms/services/submissions.client.factory.js:4]: i think we can use the full name now, dont need to rename it anymore :P

File: src/public/modules/forms/services/submissions.client.factory.js
Comment (karrui) [inline src/public/modules/forms/services/submissions.client.factory.js:186]: should we rename experimentalCsvGenerator now that it has already been in the application for a while

File: src/public/modules/forms/helpers/response-factory.ts
Comment (karrui) [inline src/public/modules/forms/helpers/response-factory.ts:68]: i think there may be issues with this check; are we certain that all answers must be string? What about number fields? They might be of `number` type instead?

File: src/public/modules/forms/helpers/response-factory.ts
Comment (karrui) [inline src/public/modules/forms/helpers/response-factory.ts:61]: same here, might be too strict

File: tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js
Comment (karrui) [reply tests/unit/frontend/forms/helpers/CsvMergedHeadersGenerator.test.js:176]: we shouldn't remove this test case. Why would it fail? If it fails does it not mean the code is not working? This test is checking that indeed fields with `isHeader = true` does not get added to the `fieldIdToQuestion` map

**Review by karrui — CHANGES_REQUESTED** [review]

**Comment by karrui** [comment]
Can we also add more test cases with the additional response type checks?

File: src/public/modules/forms/helpers/response-factory.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/response-factory.ts:68]: Responses are decrypted via the `decrypt` function in the crypto SDK which returns a `DecryptedObject` which has properties `responses: FormField[]` and `verified?: Record<string, any>`. Answers in `FormField` are supposed to be of type 
```
{ answer: string; answerArray?: never }
  | { answer?: never; answerArray: string[] | string[][] }
```
`responses` and `verified` are then combined after processing the verified fields and form the csv records. So I think number field answers are also strings after decryption. I'm not sure if we should be removing the check `typeof response.answer === 'string'` because verified field answers could be of type `any`. But I think it still works because if I'm not wrong, all verified field answers are also strings at the moment.

Could you confirm if my understanding is correct and whether you think we should remove the check I mentioned?

Comment (karrui) [reply src/public/modules/forms/helpers/response-factory.ts:68]: Should be fine then, make sure to test on all form field types just in case :P

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (karrui) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:141]: missed this, but if this is private we can use the `#` modifier too

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (karrui) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:190]: and here

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (karrui) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:64]: This won't fail with non-answer fields like Headers and Statements right? The old way returned a SingleAnswer record by default but now we throw an error.

File: src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts]: Not possible due to ECMA version :(
```
ERROR in /home/travis/build/opengovsg/FormSG/src/public/modules/forms/helpers/csv-response-classes/SingleAnswerResponse.class.ts(6,3)
          TS18028: Private identifiers are only available when targeting ECMAScript 2015 and higher.
```

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:141]: We can't do this because of the ECMA version :(
```
ERROR in /home/travis/build/opengovsg/FormSG/src/public/modules/forms/helpers/csv-response-classes/SingleAnswerResponse.class.ts(6,3)
          TS18028: Private identifiers are only available when targeting ECMAScript 2015 and higher.
```

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:64]: Yup I checked, it won't fail. They have `answer=""`, so should be fine!

File: src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts
Comment (karrui) [reply src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts]: ehhhhh we are targeting ES6 (i.e. ECMAScript 2015) in our TSConfig....

Comment (karrui) [reply src/public/modules/forms/helpers/csv-response-classes/ArrayAnswerResponse.class.ts]: use Typescript's [`private`](https://www.typescriptlang.org/docs/handbook/classes.html#public-private-and-protected-modifiers) modifier then...

File: src/public/modules/forms/helpers/response-factory.ts
Comment (chowyiyin) [reply src/public/modules/forms/helpers/response-factory.ts:68]: I think the e2e tests do check the downloaded csv content for the case when all fields are added. That should be enough right?

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (karrui) [reply src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:141]: Somehow my comment didn't go through, our TSConfig is indeed targeting ES6 though... but probably not worth it to debug. 

We can use Typescript's own [`private`](https://www.typescriptlang.org/docs/handbook/2/classes.html#private) modifier in that case

File: src/public/modules/forms/helpers/response-factory.ts
Comment (karrui) [reply src/public/modules/forms/helpers/response-factory.ts:68]: yea, don't think there's a way to programatically test them correctly since we will be mocking everything anyways

**Review by karrui — APPROVED** [review]
lgtm but wait for @mantariksh's approval too

File: src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/CsvMergedHeadersGenerator.ts:13]: noticed the `{ [fieldId: string]: Response }` shape is repeated in a couple of places. we can extract that into a type for clarity

**Review by mantariksh — APPROVED** [review]

File: src/public/modules/forms/helpers/response-factory.ts
Comment (mantariksh) [inline src/public/modules/forms/helpers/response-factory.ts:27]: we can use the enums here - `BasicField.Table`, `BasicField.Checkbox`

**Pull request merged**
